// @hubspot/local-dev-lib/api/developerTestAccounts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDeveloperTestAccountPersonalAccessKey = exports.fetchDeveloperTestAccountGateSyncStatus = exports.fetchDeveloperTestAccountOauthAppInstallStatus = exports.installOauthAppIntoDeveloperTestAccount = exports.fetchDeveloperTestAccountData = exports.deleteDeveloperTestAccount = exports.createDeveloperTestAccount = exports.fetchDeveloperTestAccounts = void 0;
const client_1 = require("../http/client");
const http_1 = require("../http");
const getAxiosConfig_1 = require("../http/getAxiosConfig");
const environments_1 = require("../constants/environments");
const api_1 = require("../constants/api");
const TEST_ACCOUNTS_API_PATH = 'integrators/test-portals/v2';
const TEST_ACCOUNTS_API_PATH_V3 = 'integrators/test-portals/v3';
function fetchDeveloperTestAccounts(accountId) {
    return http_1.http.get(accountId, {
        url: TEST_ACCOUNTS_API_PATH,
    });
}
exports.fetchDeveloperTestAccounts = fetchDeveloperTestAccounts;
function createDeveloperTestAccount(accountId, accountInfo) {
    if (typeof accountInfo === 'object') {
        return http_1.http.post(accountId, {
            url: TEST_ACCOUNTS_API_PATH_V3,
            data: accountInfo,
            timeout: api_1.SANDBOX_TIMEOUT,
        });
    }
    return http_1.http.post(accountId, {
        url: TEST_ACCOUNTS_API_PATH,
        data: { accountName: accountInfo, generatePersonalAccessKey: true },
        timeout: api_1.SANDBOX_TIMEOUT,
    });
}
exports.createDeveloperTestAccount = createDeveloperTestAccount;
function deleteDeveloperTestAccount(accountId, testAccountId, useV3 = false) {
    if (useV3) {
        return http_1.http.delete(accountId, {
            url: `${TEST_ACCOUNTS_API_PATH_V3}/${testAccountId}`,
        });
    }
    return http_1.http.delete(accountId, {
        url: `${TEST_ACCOUNTS_API_PATH}/${testAccountId}`,
    });
}
exports.deleteDeveloperTestAccount = deleteDeveloperTestAccount;
function fetchDeveloperTestAccountData(accessToken, accountId, env = environments_1.ENVIRONMENTS.PROD) {
    const axiosConfig = (0, getAxiosConfig_1.getAxiosConfig)({
        env,
        url: `${TEST_ACCOUNTS_API_PATH}/self`,
        params: { portalId: accountId },
    });
    const reqWithToken = {
        ...axiosConfig,
        headers: {
            ...axiosConfig.headers,
            Authorization: `Bearer ${accessToken}`,
        },
    };
    return (0, client_1.httpClient)(reqWithToken);
}
exports.fetchDeveloperTestAccountData = fetchDeveloperTestAccountData;
function installOauthAppIntoDeveloperTestAccount(accountId, testAccountId, projectName, appUId) {
    return http_1.http.post(accountId, {
        url: `${TEST_ACCOUNTS_API_PATH_V3}/install-apps`,
        data: {
            testPortalId: testAccountId,
            developerQualifiedSymbol: { developerSymbol: appUId, projectName },
        },
        timeout: api_1.SANDBOX_TIMEOUT,
    });
}
exports.installOauthAppIntoDeveloperTestAccount = installOauthAppIntoDeveloperTestAccount;
function fetchDeveloperTestAccountOauthAppInstallStatus(accountId, projectName, appUId) {
    return http_1.http.post(accountId, {
        url: `${TEST_ACCOUNTS_API_PATH_V3}/install-apps/readiness`,
        data: {
            developerQualifiedSymbol: { developerSymbol: appUId, projectName },
        },
        timeout: api_1.SANDBOX_TIMEOUT,
    });
}
exports.fetchDeveloperTestAccountOauthAppInstallStatus = fetchDeveloperTestAccountOauthAppInstallStatus;
function fetchDeveloperTestAccountGateSyncStatus(accountId, testAccountId) {
    return http_1.http.get(accountId, {
        url: `${TEST_ACCOUNTS_API_PATH_V3}/gate-sync-status/${testAccountId}`,
    });
}
exports.fetchDeveloperTestAccountGateSyncStatus = fetchDeveloperTestAccountGateSyncStatus;
function generateDeveloperTestAccountPersonalAccessKey(accountId, testAccountId) {
    return http_1.http.get(accountId, {
        url: `${TEST_ACCOUNTS_API_PATH_V3}/generate-pak/${testAccountId}`,
    });
}
exports.generateDeveloperTestAccountPersonalAccessKey = generateDeveloperTestAccountPersonalAccessKey;
