// @hubspot/local-dev-lib/api/github.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchRepoContents = exports.fetchRepoFileByDownloadUrl = exports.fetchRepoFile = exports.fetchRepoAsZip = exports.fetchRepoReleaseData = void 0;
const axios_1 = __importDefault(require("axios"));
const getAxiosConfig_1 = require("../http/getAxiosConfig");
const errors_1 = require("../errors");
const GITHUB_REPOS_API = 'https://api.github.com/repos';
const GITHUB_RAW_CONTENT_API_PATH = 'https://raw.githubusercontent.com';
function getAdditionalHeaders() {
    const headers = {};
    if (global && global.githubToken) {
        headers.authorization = `Bearer ${global.githubToken}`;
    }
    else if (process.env.GITHUB_TOKEN) {
        headers.authorization = `Bearer ${process.env.GITHUB_TOKEN}`;
    }
    return headers;
}
function githubRequestWithFallback(url, responseType) {
    const headersWithAuth = {
        ...(0, getAxiosConfig_1.getDefaultUserAgentHeader)(),
        ...getAdditionalHeaders(),
    };
    if (headersWithAuth.authorization) {
        return axios_1.default
            .get(url, { headers: headersWithAuth, responseType })
            .catch(error => {
            // 404 with an auth token might mean an SSO issue so retry without the authorization header
            if ((0, errors_1.isSpecifiedError)(error, { statusCode: 404 })) {
                return axios_1.default.get(url, {
                    headers: { ...(0, getAxiosConfig_1.getDefaultUserAgentHeader)() },
                    responseType,
                });
            }
            throw error;
        });
    }
    // No auth token, proceed normally
    return axios_1.default.get(url, { headers: headersWithAuth, responseType });
}
// Returns information about the repo's releases. Defaults to "latest" if no tag is provided
// https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-a-release-by-tag-name
function fetchRepoReleaseData(repoPath, tag = '') {
    const URL = `${GITHUB_REPOS_API}/${repoPath}/releases`;
    return githubRequestWithFallback(`${URL}/${tag ? `tags/${tag}` : 'latest'}`);
}
exports.fetchRepoReleaseData = fetchRepoReleaseData;
// Returns the entire repo content as a zip, using the zipball_url from fetchRepoReleaseData()
// https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#download-a-repository-archive-zip
function fetchRepoAsZip(zipUrl) {
    return axios_1.default.get(zipUrl, {
        responseType: 'arraybuffer',
        headers: { ...(0, getAxiosConfig_1.getDefaultUserAgentHeader)(), ...getAdditionalHeaders() },
    });
}
exports.fetchRepoAsZip = fetchRepoAsZip;
// Returns the raw file contents via the raw.githubusercontent endpoint
function fetchRepoFile(repoPath, filePath, ref) {
    const url = `${GITHUB_RAW_CONTENT_API_PATH}/${repoPath}/${ref}/${filePath}`;
    return githubRequestWithFallback(url);
}
exports.fetchRepoFile = fetchRepoFile;
// Returns the raw file contents via the raw.githubusercontent endpoint
function fetchRepoFileByDownloadUrl(downloadUrl) {
    return githubRequestWithFallback(downloadUrl, 'arraybuffer');
}
exports.fetchRepoFileByDownloadUrl = fetchRepoFileByDownloadUrl;
// Returns the contents of a file or directory in a repository by path
// https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content
function fetchRepoContents(repoPath, path, ref) {
    const refQuery = ref ? `?ref=${ref}` : '';
    return githubRequestWithFallback(`${GITHUB_REPOS_API}/${repoPath}/contents/${path}${refQuery}`);
}
exports.fetchRepoContents = fetchRepoContents;
