// @hubspot/local-dev-lib/http/client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpClient = void 0;
const axios_1 = __importStar(require("axios"));
const HubSpotHttpError_1 = require("../models/HubSpotHttpError");
const logger_1 = require("../lib/logger");
const localDevAuth_1 = require("../api/localDevAuth");
const fireAlarm_1 = require("../api/fireAlarm");
const util = __importStar(require("util"));
const trackUsage_1 = require("../lib/trackUsage");
// Create an isolated axios instance for this copy of local-dev-lib.
// This prevents issues when multiple copies are loaded and share the global
// axios, where each copy would register interceptors on the shared instance
// causing errors to be wrapped multiple times.
exports.httpClient = axios_1.default.create();
const IGNORE_URLS_NETWORK_DEBUG = [
    localDevAuth_1.LOCALDEVAUTH_ACCESS_TOKEN_PATH,
    trackUsage_1.CMS_CLI_USAGE_PATH,
    trackUsage_1.VSCODE_USAGE_PATH,
    fireAlarm_1.FIREALARM_API_AUTH_PATH,
];
function logRequest(response) {
    try {
        if (!process.env.HUBSPOT_NETWORK_LOGGING) {
            return;
        }
        if (!process.env.HUBSPOT_DEBUG_LOGGING_VERBOSE &&
            IGNORE_URLS_NETWORK_DEBUG.some(url => response?.config?.url && response.config.url.includes(url))) {
            return;
        }
        logger_1.logger.debug(util.inspect({
            method: response.config.method,
            baseURL: response.config.baseURL,
            url: response.config.url,
            data: response.data,
            status: response.status,
        }, false, null, true));
    }
    catch (error) {
        // Ignore any errors that occur while logging the response
    }
}
// Register interceptor on our isolated instance
exports.httpClient.interceptors.response.use((response) => {
    logRequest(response);
    return response;
}, error => {
    try {
        if ((0, axios_1.isAxiosError)(error) && error.response) {
            logRequest(error.response);
        }
    }
    catch (e) {
        // Ignore any errors that occur while logging the response
    }
    // Wrap all axios errors in our own Error class. Attach the error
    // as the cause for the new error, so we maintain the stack trace
    return Promise.reject(new HubSpotHttpError_1.HubSpotHttpError(error.message, { cause: error }));
});
