// @hubspot/local-dev-lib/api/devSecrets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchAppSecrets = exports.deleteAppSecret = exports.updateAppSecret = exports.addAppSecret = void 0;
const http_1 = require("../http");
const DEV_SECRETS_API_PATH = 'dev-secrets/management/v3';
function addAppSecret(accountId, appId, key, value) {
    return http_1.http.post(accountId, {
        url: `${DEV_SECRETS_API_PATH}/secrets/app/${appId}`,
        data: {
            key,
            value,
        },
    });
}
exports.addAppSecret = addAppSecret;
function updateAppSecret(accountId, appId, key, value) {
    return http_1.http.patch(accountId, {
        url: `${DEV_SECRETS_API_PATH}/secrets/app/${appId}`,
        data: {
            key,
            value,
        },
    });
}
exports.updateAppSecret = updateAppSecret;
function deleteAppSecret(accountId, appId, key) {
    return http_1.http.delete(accountId, {
        url: `${DEV_SECRETS_API_PATH}/secrets/app/${appId}`,
        data: {
            key,
        },
    });
}
exports.deleteAppSecret = deleteAppSecret;
function fetchAppSecrets(accountId, appId) {
    return http_1.http.get(accountId, {
        url: `${DEV_SECRETS_API_PATH}/keys/app/${appId}`,
    });
}
exports.fetchAppSecrets = fetchAppSecrets;
