// @hubspot/local-dev-lib/api/localDevAuth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchEnabledFeatures = exports.fetchAppInstallationData = exports.fetchScopeAuthorizationData = exports.fetchScopeData = exports.fetchAccessToken = exports.LOCALDEVAUTH_ACCESS_TOKEN_PATH = void 0;
const getAxiosConfig_1 = require("../http/getAxiosConfig");
const http_1 = require("../http");
const environments_1 = require("../constants/environments");
const client_1 = require("../http/client");
const LOCALDEVAUTH_API_AUTH_PATH = 'localdevauth/v1/auth';
exports.LOCALDEVAUTH_ACCESS_TOKEN_PATH = `${LOCALDEVAUTH_API_AUTH_PATH}/refresh`;
function fetchAccessToken(personalAccessKey, env = environments_1.ENVIRONMENTS.PROD, portalId) {
    const axiosConfig = (0, getAxiosConfig_1.getAxiosConfig)({
        env,
        localHostOverride: true,
        url: exports.LOCALDEVAUTH_ACCESS_TOKEN_PATH,
        data: {
            encodedOAuthRefreshToken: personalAccessKey,
        },
        params: portalId ? { portalId } : {},
    });
    return (0, client_1.httpClient)({
        ...axiosConfig,
        method: 'post',
    });
}
exports.fetchAccessToken = fetchAccessToken;
function fetchScopeData(accountId, scopeGroup) {
    return http_1.http.get(accountId, {
        url: `${LOCALDEVAUTH_API_AUTH_PATH}/check-scopes`,
        params: { scopeGroup },
    });
}
exports.fetchScopeData = fetchScopeData;
async function fetchScopeAuthorizationData(accountId) {
    return http_1.http.get(accountId, {
        url: `${LOCALDEVAUTH_API_AUTH_PATH}/scope-groups/authorized`,
    });
}
exports.fetchScopeAuthorizationData = fetchScopeAuthorizationData;
function fetchAppInstallationData(portalId, projectId, appUid, requiredScopeGroups, optionalScopeGroups = []) {
    return http_1.http.post(portalId, {
        url: `${LOCALDEVAUTH_API_AUTH_PATH}/install-info`,
        data: {
            portalId,
            projectId,
            sourceId: appUid,
            requiredScopeGroups,
            optionalScopeGroups,
        },
    });
}
exports.fetchAppInstallationData = fetchAppInstallationData;
async function fetchEnabledFeatures(accountId) {
    return http_1.http.get(accountId, {
        url: `${LOCALDEVAUTH_API_AUTH_PATH}/enabled-features`,
    });
}
exports.fetchEnabledFeatures = fetchEnabledFeatures;
