// @hubspot/local-dev-lib/constants/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HUBSPOT_CONFIG_OPERATIONS = exports.HUBSPOT_CONFIG_ERROR_TYPES = exports.ACCOUNT_IDENTIFIERS = exports.ENVIRONMENT_VARIABLES = exports.CONFIG_FLAGS = exports.HUBSPOT_ACCOUNT_TYPE_STRINGS = exports.HUBSPOT_ACCOUNT_TYPES = exports.MCP_TOTAL_TOOL_CALLS_STATE = exports.MIN_HTTP_TIMEOUT = exports.DEFAULT_PORTAL = exports.DEFAULT_ACCOUNT = exports.ALLOW_AUTO_UPDATES = exports.AUTO_OPEN_BROWSER = exports.ALLOW_USAGE_TRACKING = exports.HTTP_USE_LOCALHOST = exports.ENV = exports.HTTP_TIMEOUT = exports.DEFAULT_CMS_PUBLISH_MODE = exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND = exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID = exports.DEFAULT_ACCOUNT_OVERRIDE_FILE_NAME = exports.STATE_FILE_PATH = exports.GLOBAL_CONFIG_PATH = exports.HUBSPOT_STATE_FILE = exports.HUBSPOT_CONFIGURATION_FILE = exports.HUBSPOT_CONFIGURATION_FOLDER = exports.ARCHIVED_HUBSPOT_CONFIG_YAML_FILE_NAME = exports.DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME = void 0;
const lang_1 = require("../utils/lang");
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
exports.DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME = 'hubspot.config.yml';
exports.ARCHIVED_HUBSPOT_CONFIG_YAML_FILE_NAME = 'archived.hubspot.config.yml';
exports.HUBSPOT_CONFIGURATION_FOLDER = '.hscli';
exports.HUBSPOT_CONFIGURATION_FILE = 'config.yml';
exports.HUBSPOT_STATE_FILE = 'state.json';
exports.GLOBAL_CONFIG_PATH = path_1.default.join(os_1.default.homedir(), exports.HUBSPOT_CONFIGURATION_FOLDER, exports.HUBSPOT_CONFIGURATION_FILE);
exports.STATE_FILE_PATH = path_1.default.join(os_1.default.homedir(), exports.HUBSPOT_CONFIGURATION_FOLDER, exports.HUBSPOT_STATE_FILE);
exports.DEFAULT_ACCOUNT_OVERRIDE_FILE_NAME = '.hsaccount';
exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID = 'DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID';
exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND = 'DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND';
exports.DEFAULT_CMS_PUBLISH_MODE = 'defaultCmsPublishMode';
exports.HTTP_TIMEOUT = 'httpTimeout';
exports.ENV = 'env';
exports.HTTP_USE_LOCALHOST = 'httpUseLocalhost';
exports.ALLOW_USAGE_TRACKING = 'allowUsageTracking';
exports.AUTO_OPEN_BROWSER = 'autoOpenBrowser';
exports.ALLOW_AUTO_UPDATES = 'allowAutoUpdates';
exports.DEFAULT_ACCOUNT = 'defaultAccount';
exports.DEFAULT_PORTAL = 'defaultPortal';
exports.MIN_HTTP_TIMEOUT = 3000;
exports.MCP_TOTAL_TOOL_CALLS_STATE = 'mcpTotalToolCalls';
exports.HUBSPOT_ACCOUNT_TYPES = {
    DEVELOPMENT_SANDBOX: 'DEVELOPMENT_SANDBOX',
    DEVELOPER_TEST: 'DEVELOPER_TEST',
    APP_DEVELOPER: 'APP_DEVELOPER',
    STANDARD_SANDBOX: 'STANDARD_SANDBOX',
    STANDARD: 'STANDARD',
};
exports.HUBSPOT_ACCOUNT_TYPE_STRINGS = {
    DEVELOPMENT_SANDBOX: (0, lang_1.i18n)('lib.accountTypes.developmentSandbox'),
    STANDARD_SANDBOX: (0, lang_1.i18n)('lib.accountTypes.standardSandbox'),
    DEVELOPER_TEST: (0, lang_1.i18n)('lib.accountTypes.developerTest'),
    APP_DEVELOPER: (0, lang_1.i18n)('lib.accountTypes.appDeveloper'),
    STANDARD: (0, lang_1.i18n)('lib.accountTypes.standard'),
};
exports.CONFIG_FLAGS = {
    USE_CUSTOM_OBJECT_HUBFILE: 'useCustomObjectHubfile',
    HTTP_USE_LOCALHOST: 'httpUseLocalhost',
};
exports.ENVIRONMENT_VARIABLES = {
    HUBSPOT_API_KEY: 'HUBSPOT_API_KEY',
    HUBSPOT_CLIENT_ID: 'HUBSPOT_CLIENT_ID',
    HUBSPOT_CLIENT_SECRET: 'HUBSPOT_CLIENT_SECRET',
    HUBSPOT_PERSONAL_ACCESS_KEY: 'HUBSPOT_PERSONAL_ACCESS_KEY',
    HUBSPOT_ACCOUNT_ID: 'HUBSPOT_ACCOUNT_ID',
    HUBSPOT_PORTAL_ID: 'HUBSPOT_PORTAL_ID',
    HUBSPOT_REFRESH_TOKEN: 'HUBSPOT_REFRESH_TOKEN',
    HUBSPOT_ENVIRONMENT: 'HUBSPOT_ENVIRONMENT',
    HTTP_TIMEOUT: 'HTTP_TIMEOUT',
    HTTP_USE_LOCALHOST: 'HTTP_USE_LOCALHOST',
    ALLOW_USAGE_TRACKING: 'ALLOW_USAGE_TRACKING',
    DEFAULT_CMS_PUBLISH_MODE: 'DEFUALT_CMS_PUBLISH_MODE',
    USE_ENVIRONMENT_HUBSPOT_CONFIG: 'USE_ENVIRONMENT_HUBSPOT_CONFIG',
    HUBSPOT_CONFIG_PATH: 'HUBSPOT_CONFIG_PATH',
};
exports.ACCOUNT_IDENTIFIERS = {
    ACCOUNT_ID: 'accountId',
    NAME: 'name',
};
exports.HUBSPOT_CONFIG_ERROR_TYPES = {
    CONFIG_NOT_FOUND: 'CONFIG_NOT_FOUND',
    INSUFFICIENT_PERMISSIONS: 'INSUFFICIENT_PERMISSIONS',
    ACCOUNT_NOT_FOUND: 'ACCOUNT_NOT_FOUND',
    NO_DEFAULT_ACCOUNT: 'NO_DEFAULT_ACCOUNT',
    INVALID_ENVIRONMENT_VARIABLES: 'ENVIRONMENT_VARIABLES',
    YAML_PARSING: 'YAML_PARSING',
    INVALID_ACCOUNT: 'INVALID_ACCOUNT',
    INVALID_FIELD: 'INVALID_FIELD',
    UNKNOWN: 'UNKNOWN',
};
exports.HUBSPOT_CONFIG_OPERATIONS = {
    READ: 'READ',
    WRITE: 'WRITE',
    DELETE: 'DELETE',
};
