// @hubspot/local-dev-lib/config/migrate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.archiveConfigAtPath = exports.mergeConfigAccounts = exports.mergeConfigProperties = exports.migrateConfigAtPath = exports.getConfigAtPath = void 0;
const fs_1 = __importDefault(require("fs"));
const index_1 = require("./index");
const config_1 = require("../constants/config");
const utils_1 = require("./utils");
const path_1 = __importDefault(require("path"));
function getConfigAtPath(path) {
    const configFileSource = (0, utils_1.readConfigFile)(path);
    return (0, utils_1.parseConfig)(configFileSource, path);
}
exports.getConfigAtPath = getConfigAtPath;
function migrateConfigAtPath(path) {
    (0, index_1.createEmptyConfigFile)(true);
    const configToMigrate = getConfigAtPath(path);
    (0, utils_1.writeConfigFile)(configToMigrate, (0, index_1.getGlobalConfigFilePath)());
}
exports.migrateConfigAtPath = migrateConfigAtPath;
function mergeConfigProperties(toConfig, fromConfig, force) {
    const conflicts = [];
    if (force) {
        toConfig.defaultCmsPublishMode = fromConfig.defaultCmsPublishMode;
        toConfig.httpTimeout = fromConfig.httpTimeout;
        toConfig.env = fromConfig.env;
        toConfig.httpUseLocalhost = fromConfig.httpUseLocalhost;
        toConfig.allowUsageTracking = fromConfig.allowUsageTracking;
        toConfig.autoOpenBrowser = fromConfig.autoOpenBrowser;
        toConfig.allowAutoUpdates = fromConfig.allowAutoUpdates;
        toConfig.defaultAccount = fromConfig.defaultAccount;
    }
    else {
        toConfig.defaultCmsPublishMode ||= fromConfig.defaultCmsPublishMode;
        toConfig.httpTimeout ||= fromConfig.httpTimeout;
        toConfig.env ||= fromConfig.env;
        toConfig.httpUseLocalhost =
            toConfig.httpUseLocalhost === undefined
                ? fromConfig.httpUseLocalhost
                : toConfig.httpUseLocalhost;
        toConfig.allowUsageTracking =
            toConfig.allowUsageTracking === undefined
                ? fromConfig.allowUsageTracking
                : toConfig.allowUsageTracking;
        toConfig.autoOpenBrowser =
            toConfig.autoOpenBrowser === undefined
                ? fromConfig.autoOpenBrowser
                : toConfig.autoOpenBrowser;
        toConfig.allowAutoUpdates =
            toConfig.allowAutoUpdates === undefined
                ? fromConfig.allowAutoUpdates
                : toConfig.allowAutoUpdates;
        toConfig.defaultAccount ||= fromConfig.defaultAccount;
        const propertiesToCheck = [
            config_1.DEFAULT_CMS_PUBLISH_MODE,
            config_1.HTTP_TIMEOUT,
            config_1.ENV,
            config_1.HTTP_USE_LOCALHOST,
            config_1.ALLOW_USAGE_TRACKING,
            config_1.AUTO_OPEN_BROWSER,
            config_1.ALLOW_AUTO_UPDATES,
            config_1.DEFAULT_ACCOUNT,
        ];
        propertiesToCheck.forEach(prop => {
            if (toConfig[prop] !== undefined &&
                fromConfig[prop] !== undefined &&
                toConfig[prop] !== fromConfig[prop]) {
                conflicts.push({
                    property: prop,
                    oldValue: fromConfig[prop],
                    newValue: toConfig[prop],
                });
            }
        });
    }
    // Merge flags
    if (toConfig.flags || fromConfig.flags) {
        toConfig.flags = Array.from(new Set([...(toConfig.flags || []), ...(fromConfig.flags || [])]));
    }
    return { configWithMergedProperties: toConfig, conflicts };
}
exports.mergeConfigProperties = mergeConfigProperties;
function buildConfigWithMergedAccounts(toConfig, fromConfig) {
    const existingAccountIds = toConfig.accounts.map(({ accountId }) => accountId);
    const skippedAccountIds = [];
    fromConfig.accounts.forEach(account => {
        if (existingAccountIds.includes(account.accountId)) {
            skippedAccountIds.push(account.accountId);
        }
        else {
            toConfig.accounts.push(account);
        }
    });
    return {
        configWithMergedAccounts: toConfig,
        skippedAccountIds,
    };
}
function mergeConfigAccounts(toConfig, fromConfig) {
    const { configWithMergedAccounts, skippedAccountIds } = buildConfigWithMergedAccounts(toConfig, fromConfig);
    (0, utils_1.writeConfigFile)(configWithMergedAccounts, (0, index_1.getGlobalConfigFilePath)());
    return { configWithMergedAccounts, skippedAccountIds };
}
exports.mergeConfigAccounts = mergeConfigAccounts;
function archiveConfigAtPath(configPath) {
    const dir = path_1.default.dirname(configPath);
    const archivedConfigPath = path_1.default.join(dir, config_1.ARCHIVED_HUBSPOT_CONFIG_YAML_FILE_NAME);
    fs_1.default.renameSync(configPath, archivedConfigPath);
}
exports.archiveConfigAtPath = archiveConfigAtPath;
