// @hubspot/local-dev-lib/lib/trackUsage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackUsage = exports.VSCODE_USAGE_PATH = exports.CMS_CLI_USAGE_PATH = void 0;
const client_1 = require("../http/client");
const getAxiosConfig_1 = require("../http/getAxiosConfig");
const logger_1 = require("./logger");
const http_1 = require("../http");
const config_1 = require("../config");
const fileMapper_1 = require("../api/fileMapper");
const lang_1 = require("../utils/lang");
const environment_1 = require("./environment");
const i18nKey = 'lib.trackUsage';
exports.CMS_CLI_USAGE_PATH = `${fileMapper_1.FILE_MAPPER_API_PATH}/cms-cli-usage`;
exports.VSCODE_USAGE_PATH = `${fileMapper_1.FILE_MAPPER_API_PATH}/vscode-extension-usage`;
async function trackUsage(eventName, eventClass, meta = {}, accountId) {
    const usageEvent = {
        accountId,
        eventName,
        eventClass,
        meta,
    };
    const EVENT_TYPES = {
        VSCODE_EXTENSION_INTERACTION: 'vscode-extension-interaction',
        CLI_INTERACTION: 'cli-interaction',
    };
    let path = fileMapper_1.FILE_MAPPER_API_PATH;
    switch (eventName) {
        case EVENT_TYPES.CLI_INTERACTION:
            path = exports.CMS_CLI_USAGE_PATH;
            break;
        case EVENT_TYPES.VSCODE_EXTENSION_INTERACTION:
            path = exports.VSCODE_USAGE_PATH;
            break;
        default:
            logger_1.logger.debug((0, lang_1.i18n)(`${i18nKey}.invalidEvent`, { eventName }));
    }
    const account = accountId && (0, config_1.getConfigAccountById)(accountId);
    if (account && account.authType === 'personalaccesskey') {
        logger_1.logger.debug((0, lang_1.i18n)(`${i18nKey}.sendingEventAuthenticated`));
        try {
            await http_1.http.post(accountId, {
                url: `${path}/authenticated`,
                data: usageEvent,
                resolveWithFullResponse: true,
            });
            return;
        }
        catch (e) {
            logger_1.logger.debug((0, lang_1.i18n)(`${i18nKey}.retryingEventUnauthenticated`));
        }
    }
    const env = accountId
        ? (0, config_1.getConfigAccountEnvironment)(accountId)
        : (0, environment_1.getValidEnv)();
    const axiosConfig = (0, getAxiosConfig_1.getAxiosConfig)({
        env,
        url: path,
        data: usageEvent,
        resolveWithFullResponse: true,
    });
    logger_1.logger.debug((0, lang_1.i18n)(`${i18nKey}.sendingEventUnauthenticated`));
    return (0, client_1.httpClient)({ ...axiosConfig, method: 'post' });
}
exports.trackUsage = trackUsage;
